core/solver/batch_dispatch.hpp.inc

```diff
@@ -0,0 +1,181 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+template <typename DeviceValueType>
+class DummyKernelCaller {
+public:
+    template <typename BatchMatrixType, typename PrecType, typename StopType,
+              typename LogType>
+    void call_kernel(
+        LogType logger, const BatchMatrixType& a,
+        const gko::batch_dense::UniformBatch<DeviceValueType>& b,
+        const gko::batch_dense::UniformBatch<DeviceValueType>& x) const
+    {}
+};
+
+
+/**
+ * Handles dispatching to the correct instantiation of a batched solver
+ * depending on runtime parameters.
+ *
+ * @tparam KernelCaller  Class with an interface like DummyKernelCaller,
+ *   that is reponsible for finally calling the templated backend-specific
+ *   kernel.
+ * @tparam ExecType  Executor to run the kernel on.
+ * @tparam OptsType  Structure type of options for the particular solver to be
+ *   used.
+ * @tparam ValueType  The user-facing value type.
+ * @tparam DeviceValueType  The backend-specific value type corresponding to
+ *   ValueType.
+ */
+template <typename KernelCaller, typename ExecType, typename OptsType,
+          typename ValueType, typename DeviceValueType>
```

Comment:
Why have both ValueType and DeviceValueType, I think there is a one to one correspondence, so you should be able to cast one to the other and there should be no need for templating ? 

I think a single ValueType makes it easy for mixed precision inference later as well.

Reply:
I'll try to think about how that can be done. The issue is this same code is used for all backends, so that correspondence depends on the backend.

Reply:
This is now addressed.